thewalrus/csamples.py

```diff
@@ -0,0 +1,142 @@
+
+Summary
+-------
+.. autosummary::
+    rescale_adjacency_matrix_thermal
+    rescale_adjacency_matrix
+    generate_thermal_samples
+
+Code details
+------------
+"""
+# pylint: disable=too-many-arguments
+
+import numpy as np
+from scipy.optimize import root_scalar
+
+def rescale_adjacency_matrix_thermal(
+    A, n_mean, check_positivity=True, check_symmetry=True, rtol=1e-05, atol=1e-08
+):
+    r""" Returns the scaling parameter by which the adjacency matrix A
+    should be rescaled so that the Gaussian state that encodes it has
+    a total mean photon number n_mean for thermal sampling.
+    Args:
+        A (array): Adjacency matrix, assumed to be positive semi-definite and real
+        n_mean (float): Mean photon number of the Gaussian state
+        check_positivity (bool): Checks if the matrix A is positive semidefinite
+        check_symmetry (bool): Checks if the matrix is symmetric
+        rtol: relative tolerance for the checks
+        atol: absolute tolerance for the checks
+    Returns:
+        tuple(array,array): rescaled eigenvalues and eigenvectors of the matrix A
+    """
+    return rescale_adjacency_matrix(
+        A, n_mean, 1.0, check_positivity=check_positivity, check_symmetry=check_symmetry, rtol=rtol, atol=atol
+    )
+
+def rescale_adjacency_matrix(
+    A, n_mean, scale, check_positivity=True, check_symmetry=True, rtol=1e-05, atol=1e-08
+):
+    r""" Returns the scaling parameter by which the adjacency matrix A
+    should be rescaled so that the Gaussian state that encodes it has
+    a total mean photon number n_mean.
+    Args:
+        A (array): Adjacency matrix, assumed to be positive semi-definite and real
+        n_mean (float): Mean photon number of the Gaussian state
+        scale (float): Determines whether to rescale the matrix for thermal sampling (scale = 1.0)
+            or for squashed sampling (scale = 2.0)
+        check_positivity (bool): Checks if the matrix A is positive semidefinite
+        check_symmetry (bool): Checks if the matrix is symmetric
+        rtol: relative tolerance for the checks
+        atol: absolute tolerance for the checks
+    Returns:
+        tuple(array,array): rescaled eigenvalues and eigenvectors of the matrix A
+    """
+    ls, O = np.linalg.eigh(A)
+    ls[np.abs(ls) < atol] = 0.0
+
+    if check_symmetry is True:
+        assert np.allclose(A, A.T, rtol=rtol, atol=atol)
+
+    if check_positivity is True:
+        assert np.all(ls >= 0)
+
+    max_sv = ls[-1]
+    a_lim = 0.0
+    b_lim = 1 / (atol + scale * max_sv)
+    x_init = 0.5 * b_lim
+
+    def mean_photon_number(x, vals):
+        r""" Returns the mean number of photons in the Gaussian state that
+        encodes the adjacency matrix x*A where vals are the eigenvalues of positive semidefinite A
+        Args:
+            x (float): Scaling parameter
+            vals (array): Eigenvalues of the matrix A
+        Returns:
+            n_mean: Mean photon number in the Gaussian state
```

Comment:
The first thing in the returns line should be the _type_ of thing returned rather than its name